pkg/publisher/util/utils.go

```diff
@@ -23,35 +24,49 @@ func NewPublisherProvider(
 	ctx context.Context,
 	storagePath string,
 	cm *system.CleanupManager,
-	ipfsConnect string,
+	cfg types2.PublishersConfig,
 	localConfig *types.LocalPublisherConfig,
 ) (publisher.PublisherProvider, error) {
-	noopPublisher := noop.NewNoopPublisher()
-	s3Publisher, err := configureS3Publisher(storagePath, cm)
-	if err != nil {
-		return nil, err
-	}
+	providers := make(map[string]publisher.Publisher)
 
-	localPublisher := local.NewLocalPublisher(ctx, localConfig.Directory, localConfig.Address, localConfig.Port)
+	// TODO(review): the NoopPublisher is a testing attribute, should it remain part of the default set of publishers?
+	// for now, lets keep everything the same for compatibility, else tests might fail.
+	providers[models.PublisherNoop] = noop.NewNoopPublisher()
 
-	providers := map[string]publisher.Publisher{
-		models.PublisherNoop:  tracing.Wrap(noopPublisher),
-		models.PublisherS3:    tracing.Wrap(s3Publisher),
-		models.PublisherLocal: tracing.Wrap(localPublisher),
+	if cfg.Enabled(types2.KindPublisherS3) {
+		s3Publisher, err := configureS3Publisher(storagePath, cm)
+		if err != nil {
+			return nil, err
+		}
+		providers[models.PublisherS3] = tracing.Wrap(s3Publisher)
 	}
 
-	if ipfsConnect != "" {
-		ipfsClient, err := ipfs_client.NewClient(ctx, ipfsConnect)
+	if cfg.Enabled(types2.KindPublisherLocal) {
+		providers[models.PublisherLocal] = tracing.Wrap(local.NewLocalPublisher(
+			ctx,
+			localConfig.Directory,
+			localConfig.Address,
+			localConfig.Port,
+		))
+	}
+
+	// TODO(review): what does it mean if IPFS isn't disabled in the config, and also doesn't have a config?
+	if cfg.Enabled(types2.KindPublisherIPFS) && cfg.HasConfig(types2.KindPublisherIPFS) {
+		ipfscfg, err := types2.DecodeProviderConfig[types2.IpfsPublisherConfig](cfg)
+		if err != nil {
+			return nil, err
+		}
+		ipfsClient, err := ipfs_client.NewClient(ctx, ipfscfg.Connect)
 		if err != nil {
 			return nil, err
 		}
 		ipfsPublisher, err := ipfs.NewIPFSPublisher(ctx, *ipfsClient)
 		if err != nil {
 			return nil, err
 		}
-
-		providers[models.PublisherIPFS] = ipfsPublisher
+		providers[models.PublisherIPFS] = tracing.Wrap(ipfsPublisher)
```

Comment:
@wdbaruni can you help me understand how to address this TODO? I assume the IPFS publisher is a (somewhat) special case and its lack of presence in the `Disabled` field doesn't imply it's enabled since configuration must be present to use it?

Reply:
the logic is similar to today's behaviour. A publisher needs to be installed and not explicitly disabled for it to be functional. In this case, if no ipfs endpoint is provided, then it is not installed. The same for s3 where credetials must be provided or available for it to be installed. Also the same with docker where the daemon must be installed locally.

The behaviour is to auto-enable publishers, engines and inputs if they are installed, unless the user explicitly opts out and disables them